### PR TITLE
tooling: Add docker build util

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:

adds a util to build docker images using `aiodocker` with a full build context

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
coverage is here https://storage.googleapis.com/envoy-pr/17384/tooling/index.html

File: tools/docker/utils.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    if log:
+        log("".join(line["stream"] for line in build if "stream" in line))
+
+    # ensure the image has been built
+    await docker.images.inspect(tag)
+
+
+async def build_image(
+        docker: aiodocker.Docker,
+        path: str,
```
Comment (phlax) [inline tools/docker/utils.py:34]: rename `path` -> `context`

**Comment by phlax** [comment]
related - i have opened a ticket upstream https://github.com/aio-libs/aiodocker/issues/602 with an offer to add a util which could simplify this code a bit

File: tools/docker/utils.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    # create a tarfile from the supplied directory
+    with tarfile.open(tar.name, fileobj=tar, mode="w") as tarball:
+        tarball.add(context, arcname=".")
+    tar.seek(0)
+
+    # build the docker image
+    build = await docker.images.build(
+        fileobj=tar, encoding="gzip", tag=tag, buildargs=buildargs or {}, **kwargs)
+    if stream:
+        stream("".join(line["stream"] for line in build if "stream" in line))
```
Comment (phlax) [inline tools/docker/utils.py:27]: if `build` is an iterator this will fail when `stream` is not set

**Comment by phlax** [comment]
WIPping this as im thinking about an implementation where its not currently required.

**Comment by phlax** [comment]
after a bit of further testing i realised this would increase the complexity as it would require build a volume for the artefacts

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**